### PR TITLE
LPS-79417 Adding test for the new social bookmark template

### PR DESCRIPTION
**Pull request by 4lejandrito**
Hi @gamerson,

I have created a template for a new kind of module (social bookmark). Everything works as expected when I create it using maven:

```
mvn archetype:generate -B -DgroupId=foo -DartifactId=com.liferay.test.bookmark -Dpackage=com.liferay.test -DarchetypeArtifactId=com.liferay.project.templates.social.bookmark -DarchetypeGroupId=com.liferay -DarchetypeVersion=1.0.0 -Dauthor=Alex -DbuildType=gradle -DclassName=Example
```

I have also written [a test](https://github.com/4lejandrito/liferay-portal/blob/61b16cc91740b88e3fb5ea512f0f77291b50e1ca/modules/sdk/project-templates/project-templates/src/test/java/com/liferay/project/templates/ProjectTemplatesTest.java#L1098) for it which works fine until it reaches [this line](https://github.com/4lejandrito/liferay-portal/blob/61b16cc91740b88e3fb5ea512f0f77291b50e1ca/modules/sdk/project-templates/project-templates/src/test/java/com/liferay/project/templates/ProjectTemplatesTest.java#L1118) throwing the following error:

```
Found differences [File /var/folders/z0/bbl10qr533j7fhwx1hj4f0qh0000gp/T/junit8376746172958853502/maven/bookmark/bnd.bnd does not match with /var/folders/z0/bbl10qr533j7fhwx1hj4f0qh0000gp/T/junit8376746172958853502/archetyper/bookmark/bnd.bnd, File /var/folders/z0/bbl10qr533j7fhwx1hj4f0qh0000gp/T/junit8376746172958853502/maven/bookmark/src/main/java/com/liferay/test/social/bookmark/FooSocialBookmark.java does not match with /var/folders/z0/bbl10qr533j7fhwx1hj4f0qh0000gp/T/junit8376746172958853502/archetyper/bookmark/src/main/java/com/liferay/test/social/bookmark/FooSocialBookmark.java, File /var/folders/z0/bbl10qr533j7fhwx1hj4f0qh0000gp/T/junit8376746172958853502/maven/bookmark/src/main/resources/content/Language.properties does not match with /var/folders/z0/bbl10qr533j7fhwx1hj4f0qh0000gp/T/junit8376746172958853502/archetyper/bookmark/src/main/resources/content/Language.properties]
```
I am still trying to understand what those changes are but... could you please have a look just in case you see which piece I am missing?

Thanks!

### CODE REVIEW

**Comment by 4lejandrito** [comment]
Hi @gamerson,

I found the issue. My template has a custom parameter [socialBookmarkType](https://github.com/4lejandrito/liferay-portal/blob/61b16cc91740b88e3fb5ea512f0f77291b50e1ca/modules/sdk/project-templates/project-templates-social-bookmark/src/main/resources/META-INF/maven/archetype-metadata.xml#L37) that I don't know how to inject inside the tests. 

The only way I found was adding a default value [inside a customizer](https://github.com/4lejandrito/liferay-portal/blob/61b16cc91740b88e3fb5ea512f0f77291b50e1ca/modules/sdk/project-templates/project-templates-social-bookmark/src/main/java/com/liferay/project/templates/social/bookmark/internal/SocialBookmarkProjectTemplateCustomizer.java#L48). This seems to be working only in one scenario but not the other:

* Inside the `maven` folder I get the default value coming from the `archetype-metadata.xml`.
* Inside the `archetyper` folder I get the default value coming from the customizer. 

If I don't use the customizer the test will complain because I am not passing a value for my parameter.

Can you please explain what the customizer is for and how to overcome this problem?

Thanks!

**Comment by gamerson** [comment]
hey @4lejandrito  thanks for this new template!  I'll review it later today and give you some feedback on how to get unstuck!

**Comment by gamerson** [comment]
Hey @4lejandrito 

So we have two ways of using these archetypes. 

1. normal maven command line `mvn archetype:generate`
2. through blade cli (which has archetypes embedded)

The 2nd way if where we use these customizers.  But actually we are going to be refactoring blade in the future to not need these `onBefore` customizer.  The `onAfter` part of the customizer is to help us with the fact that the maven archetype framework doesn't support `optional files` and for some of our archtypes we have some files that exist in 7.1 but don't exist for 7.0, etc, so we added this `customizer` interface so templates can be more complex.  This doesn't work in the maven mode (1) but that is something we are addressing in the future.

File: modules/sdk/project-templates/project-templates-social-bookmark/src/main/resources/archetype-resources/src/main/resources/META-INF/resources/page.jsp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+<%@ include file="/init.jsp" %>
+
+<%
+SocialBookmark socialBookmark = (SocialBookmark)request.getAttribute("liferay-social-bookmarks:bookmark:socialBookmark");
+String title = GetterUtil.getString((String)request.getAttribute("liferay-social-bookmarks:bookmark:title"));
+String url = GetterUtil.getString((String)request.getAttribute("liferay-social-bookmarks:bookmark:url"));
+%>
+
+<clay:link
```
Comment (gamerson) [inline modules/sdk/project-templates/project-templates-social-bookmark/src/main/resources/archetype-resources/src/main/resources/META-INF/resources/page.jsp:25]: Are clay tags only supported on liferay version 7.1?

**Comment by gamerson** [comment]
Thanks so much for all of this work! @4lejandrito   You did really well getting this far just by investigation, we have added no documentation for this :)

I'm assuming that this is only supported in 7.1?

**Comment by 4lejandrito** [comment]
Hi @gamerson,

You're correct, this is 7.1 only. That was the next thing I wanted to discuss with you after fixing the error. I guess I will need to wait until 7.1 is released to actually merge these changes?

I recorded [a videogif](http://recordit.co/BQ9hDN6r4i) for you. The steps I followed are:
* Install the template to my local repo.
* Create a fresh project using the archetype.
* Deploy the module just created.
* Go to blogs (asuming you have a blog entry already created).
* Go to settings and put the "Example" social bookmark in the top 3 so that it is displayed.
* Check that it is displayed (the magnifying lenses).

Thanks for your help!

**Comment by christopherbryanboyd** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by christopherbryanboyd** [comment]
Hi @4lejandrito , thank you for the hard work in creating this new project template!
I wanted to ask you about this line here:
https://github.com/4lejandrito/liferay-portal/blob/61b16cc91740b88e3fb5ea512f0f77291b50e1ca/modules/sdk/project-templates/project-templates-social-bookmark/src/main/resources/archetype-resources/bnd.bnd#L2
In most of our project templates, the Bundle-SymbolicName is defined as
Bundle-SymbolicName: ${package}
I see a new property was introduced into _archetype-metadata.xml_
https://github.com/4lejandrito/liferay-portal/blob/61b16cc91740b88e3fb5ea512f0f77291b50e1ca/modules/sdk/project-templates/project-templates-social-bookmark/src/main/resources/META-INF/maven/archetype-metadata.xml#L37-L39
Can you tell me about this property, why it is necessary instead of something like class name?

**Comment by 4lejandrito** [comment]
See https://github.com/christopherbryanboyd/liferay-portal/pull/11.

**Pull request closed without merging**